crates/nu-protocol/src/parse_error.rs

```diff
@@ -506,3 +511,37 @@ impl ParseError {
         }
     }
 }
+
+#[derive(Clone, Debug)]
+pub struct DidYouMean(Option<String>);
+
+fn did_you_mean_impl(possibilities_bytes: &[&[u8]], input_bytes: &[u8]) -> Option<String> {
+    let input = from_utf8(input_bytes).ok()?;
+    let possibilities = possibilities_bytes
+        .iter()
+        .map(|p| from_utf8(p))
+        .collect::<Result<Vec<&str>, Utf8Error>>()
+        .ok()?;
+    did_you_mean(&possibilities, input)
+}
+impl DidYouMean {
+    pub fn new(possibilities_bytes: &[&[u8]], input_bytes: &[u8]) -> DidYouMean {
+        DidYouMean(did_you_mean_impl(possibilities_bytes, input_bytes))
+    }
+}
+
+impl From<Option<String>> for DidYouMean {
+    fn from(value: Option<String>) -> Self {
+        Self(value)
+    }
+}
+
+impl Display for DidYouMean {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        if let Some(suggestion) = &self.0 {
+            write!(f, "Did you mean '{}'?", suggestion)
```

Comment:
I like the elegance of this, that we do not need to already store the whole string and instead can only run the formatting when the error needs to be `Display`ed.

One caveat for the variables would be that the user should get the suggestion with `$found` for ease of copy paste.

Reply:
I'm not sure I follow. Do you mean that the `$` sigil should be included in the error message?

Reply:
It depends. In the case of a mistyped `$foi` in a "dereferencing" variable it should be shown. 
Also if you forget the `$` and have a command not found error that could be made more helpful by finding that a variable matches it that needs to be indicated with `$` this part of the message would be essential.
I could imagine other uses for such a `DidYouMean` where we wouldn't want to have the `$` baked in (if it transfers to files/commands/keywords etc.)

Reply:
ok, that makes sense. I'm not sure how feasible that will end up being until I have an idea around what is going wrong with `list_variables`, but I'll take a crack at it.